### DISCUSSION TITLE
Negative sRGB values from "colour.XYZ_to_sRGB" definition.

### DISCUSSION THREAD

**Discussion by lenzls** [colour-science/colour#699] [Q&A] [answered]
Hello,

I'm trying to transform spectral intensities into RGB colors.

```
spd = colour.SpectralPowerDistribution("test", data, "test")
cmfs = colour.CMFS.get('CIE 1931 2 Degree Standard Observer')
illuminant = colour.ILLUMINANTS_RELATIVE_SPDS.get('D65')

xyz = colour.spectral_to_XYZ(spd, cmfs, illuminant)
srgb = colour.XYZ_to_sRGB(xyz/100)
```

But in the `XYZ_to_sRGB` function I always receive the following error:

```
~/.local/lib/python3.4/site-packages/colour/models/dataset/srgb.py:117: RuntimeWarning: invalid value encountered in power
  1.055 * (value ** (1 / 2.4)) - 0.055)
```

And even worse, the function gives me negative values for some input data. Thus i can't plot the color because of invalid values.
I helped myself with the following line: `srgb_corrected = [max(x, 0) for x in srgb]`, but I neither know if the RGB values are correct nor if this is a viable solution. So I'm asking here.

eg:

```
output spectral_to_XYZ:  [ 0.76537856  0.36011351  0.03784461]
output XYZ_to_sRGB:     [ 0.14769124 -0.00835964  0.001179  ]
corrected sRGB:             [0.14769123588343716, 0, 0.0011790034009353393]
```

for other datapoints it works fine:

```
output spectral_to_XYZ:  [ 44.49664248  46.29872074  50.29400567]
output XYZ_to_sRGB:     [ 0.72173567  0.70709492  0.70974093]
corrected sRGB:          [0.72173567373747338, 0.70709492126250351, 0.70974092764844987]
```

With `apply_OECF=False` the error is gone, but the values are again negative. (And I don't really know the implications of that)

And apologies if I'm doing something completely wrong. I'm not that familiar with the field and use the colour-library only for a specific task :)

Regards

Edit: 0.3.7 is my version

**Comment by thomasmansencal** [comment, marked as answer]
Hi @lensz,

Negative values are not unexpected and bad, they are actually very desirable!

Let me explain: a given RGB colourspace gamut may or may not be able to encode the entire visible spectrum, as a matter of fact, only a few do such as **ACES 2065-1**. 

When you work with 8bit or 16bit integers, what usually happens is that any colour that cannot be encoded is clipped to stay within the bounds of the RGB colourspace gamut. This effectively destroys colour information, meaning that if you wanted to go back from sRGB colourspace to CIE XYZ colourspace in your case, you wouldn't retrieve the original tristimulus values.

`Colour` operates with 32bit floating point numbers, which yield negative values in your case.

As an example, I plotted _ColorChecker N Ohta_ SPDs on the 1931 chromaticity diagram, notice how the cyan colour is out of sRGB gamut:

![image](https://cloud.githubusercontent.com/assets/99779/13550193/22211a64-e37d-11e5-91db-12528cfbc149.png)

> But in the XYZ_to_sRGB function I always receive the following error:

This is actually a warning from Numpy and can safely be ignored in that case, the sRGB OECF power function receive a value that makes it unhappy.

> Thus i can't plot the color because of invalid values.

That's where you need to choose how you want to handle those colours, they cannot be properly represented and plotted by sRGB colourspace, thus you have to alter them if you want to do so (and likely explain that they were altered).

#191 has some quite good related references about that!

**Comment by sobotka** [comment]
Technically, those negative values become non-data in the down gamut space, and as such, they can and should be clamped on the user side, I believe.

**Comment by thomasmansencal** [comment]
> they can and should be clamped on the user side, I believe.

Yeah, in the context of OP I think they should.

**Comment by thomasmansencal** [comment]
@lensz Hope the explanations were ok, I'll close that issue for now!
